core/templates/pages/exploration-editor-page/exploration-save-and-publish-buttons/exploration-save-and-publish-buttons.component.ts

```diff
@@ -129,14 +133,32 @@ angular.module('oppia').component('explorationSaveAndPublishButtons', {
         $scope.saveIsInProcess = false;
         $scope.publishIsInProcess = false;
         $scope.loadingDotsAreShown = false;
+      };
+
+      $scope.explorationCanBePublished = false;
 
+      $scope.showPublishButton = function() {
         UserExplorationPermissionsService.getPermissionsAsync()
           .then(function(permissions) {
-            $scope.showPublishButton = function() {
-              return permissions.canPublish && $scope.isPrivate();
-            };
+            $scope.explorationCanBePublished = permissions.canPublish;
+            // TODO(#8521): Remove the use of $rootScope.$apply()
+            // once the controller is migrated to angular.
+            $rootScope.$applyAsync();
           });
+
+        return $scope.explorationCanBePublished && $scope.isPrivate();
       };
+
+      $scope.$watch(function() {
```

Comment:
Please avoid adding explicit watchers! (Can you see another way to do this?)

Reply:
Fixed and also added a callback function in preventpageunloadservice. I'm not using it here